Mathlib/RingTheory/Coalgebra/GroupLike.lean

```diff
@@ -0,0 +1,155 @@
+  comul_eq_tmul_self : comul a = a ⊗ₜ[R] a
+
+attribute [simp] IsGroupLikeElem.counit_eq_one IsGroupLikeElem.comul_eq_tmul_self
+
+@[simp] lemma isGroupLikeElem_self {r : R} : IsGroupLikeElem R r ↔ r = 1 := by
+  simp +contextual [isGroupLikeElem_iff]
+
+lemma IsGroupLikeElem.ne_zero [Nontrivial R] (ha : IsGroupLikeElem R a) : a ≠ 0 := by
+  rintro rfl; simpa using ha.counit_eq_one
+
+/-- A coalgebra homomorphism sends group-like elements to group-like elements. -/
+lemma IsGroupLikeElem.map [FunLike F A B] [CoalgHomClass F R A B] (f : F)
+    (ha : IsGroupLikeElem R a) : IsGroupLikeElem R (f a) where
+  counit_eq_one := by rw [CoalgHomClass.counit_comp_apply, ha.counit_eq_one]
+  comul_eq_tmul_self := by rw [← CoalgHomClass.map_comp_comul_apply, ha.comul_eq_tmul_self]; simp
+
+/-- A coalgebra isomorphism preserves group-like elements. -/
+@[simp] lemma isGroupLikeElem_map [EquivLike F A B] [CoalgEquivClass F R A B] (f : F) :
+    IsGroupLikeElem R (f a) ↔ IsGroupLikeElem R a where
+  mp ha := by
+    rw [← (CoalgEquivClass.toCoalgEquiv f).symm_apply_apply a]
+    exact ha.map (CoalgEquivClass.toCoalgEquiv f).symm
+  mpr := .map f
+
+variable (R A) in
+/-- The type of group-like elements in a coalgebra. -/
+@[ext]
+structure GroupLike where
+  /-- The underlying element of a group-like element. -/
+  val : A
+  isGroupLikeElem_val : IsGroupLikeElem R val
+
+namespace GroupLike
+
+attribute [coe] val
+
+instance instCoeOut : CoeOut (GroupLike R A) A where coe := val
+
+attribute [simp] isGroupLikeElem_val
+
+lemma val_injective : Injective (val : GroupLike R A → A) := by rintro ⟨a, ha⟩; congr!
+
+@[simp, norm_cast] lemma val_inj {a b : GroupLike R A} : a.val = b.val ↔ a = b :=
+  val_injective.eq_iff
+
+/-- Identity equivalence between `GroupLike R A` and `{a : A | IsGroupLikeElem R a}`. -/
+def valEquiv : GroupLike R A ≃ Subtype (IsGroupLikeElem R : A → Prop) where
+  toFun a := ⟨a.1, a.2⟩
+  invFun a := ⟨a.1, a.2⟩
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+end GroupLike
+end CommSemiring
+
+section CommRing
+variable [CommRing R] [IsDomain R] [AddCommGroup A] [Module R A] [Coalgebra R A]
+  [NoZeroSMulDivisors R A]
+
+open Submodule in
+/-- Group-like elements over a domain are linearly independent. -/
+lemma linearIndepOn_isGroupLikeElem : LinearIndepOn R id {a : A | IsGroupLikeElem R a} := by
+  classical
+  -- We show that any finset `s` of group-like elements is linearly independent.
+  rw [linearIndepOn_iff_linearIndepOn_finset]
+  rintro s hs
+  -- For this, we do induction on `s`.
+  induction s using Finset.cons_induction with
+  -- The case `s = ∅` is trivial.
+  | empty => simp
+  -- Let's deal with the `s ∪ {a}` case.
+  | cons a s has ih =>
+  simp only [Finset.cons_eq_insert, Finset.coe_insert, Set.subset_def, Set.mem_insert_iff,
+    Finset.mem_coe, Set.mem_setOf_eq, forall_eq_or_imp] at hs
+  obtain ⟨ha, hs⟩ := hs
+  specialize ih hs
+  -- Assume that there is some `c : A → R` such that `∑ x ∈ s, c x • x = -c a • a`.
```

Comment:
Does it make sense to have something like [LinearIndepOn.id_insert](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/LinearIndependent/Lemmas.html#LinearIndepOn.id_insert) but for general rings so that we don't need to deal with the negative sign?

Reply:
```lean
import Mathlib

theorem LinearIndepOn.id_insert' {R M : Type*} [Ring R] [AddCommGroup M] [Module R M]
    {s : Set M} {x : M} (hs : LinearIndepOn R id s)
    (hx : ∀ r : R, r • x ∈ Submodule.span R s → r = 0) :
    LinearIndepOn R id (insert x s) :=
  sorry
```